sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1045 @@
+          null);
+    }
+
+    protected HadoopInputFormatBoundedSource(
+        SerializableConfiguration conf,
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        SimpleFunction<?, K> keyTranslationFunction,
+        SimpleFunction<?, V> valueTranslationFunction,
+        SerializableSplit inputSplit) {
+      this.conf = conf;
+      this.inputSplit = inputSplit;
+      this.keyCoder = keyCoder;
+      this.valueCoder = valueCoder;
+      this.keyTranslationFunction = keyTranslationFunction;
+      this.valueTranslationFunction = valueTranslationFunction;
+    }
+
+    public SerializableConfiguration getConfiguration() {
+      return conf;
+    }
+
+    @Override
+    public void validate() {
+      checkNotNull(conf, HadoopInputFormatIOConstants.MISSING_CONFIGURATION_SOURCE_ERROR_MSG);
+      checkNotNull(keyCoder, HadoopInputFormatIOConstants.MISSING_KEY_CODER_SOURCE_ERROR_MSG);
+      checkNotNull(valueCoder, HadoopInputFormatIOConstants.MISSING_VALUE_CODER_SOURCE_ERROR_MSG);
+    }
+
+    @Override
+    public List<BoundedSource<KV<K, V>>> splitIntoBundles(long desiredBundleSizeBytes,
+        PipelineOptions options) throws Exception {
+      if (inputSplit == null) {
+        computeSplitsIfNecessary();
+        LOG.info("Generated {} splits. Size of first split is {} ", inputSplits.size(),
+            inputSplits.get(0).getSplit().getLength());
+        return Lists.transform(inputSplits,
+            new Function<SerializableSplit, BoundedSource<KV<K, V>>>() {
+              @Override
+              public BoundedSource<KV<K, V>> apply(SerializableSplit serializableInputSplit) {
+                HadoopInputFormatBoundedSource<K, V> hifBoundedSource =
+                    new HadoopInputFormatBoundedSource<K, V>(conf, keyCoder, valueCoder,
+                        keyTranslationFunction, valueTranslationFunction,
+                        serializableInputSplit);
+                return hifBoundedSource;
+              }
+            });
+      } else {
+        LOG.info("Not splitting source {} because source is already split.", this);
+        return ImmutableList.of((BoundedSource<KV<K, V>>) this);
+      }
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions po) throws Exception {
+      if (inputSplit == null) {
+        // If there are no splits computed yet, then retrieve the splits.
+        computeSplitsIfNecessary();
+        return boundedSourceEstimatedSize;
+      }
+      return inputSplit.getSplit().getLength();
+    }
+
+    /**
+     * This is a helper function to compute splits. This method will also calculate size of the data
+     * being read. Note: This method is executed exactly once, the splits are retrieved and cached for
+     * further use by splitIntoBundles() and getEstimatesSizeBytes().
+     */
+    @VisibleForTesting
+    void computeSplitsIfNecessary() throws IOException, InterruptedException {
+      if (inputSplits == null) {
```

Comment:
same comment as above -- simplify the code and remove a layer of nesting by inverting the if and returning early.